mesh/core/src/main/protobuf/dtab.proto

```diff
@@ -0,0 +1,99 @@
+syntax = "proto3";
+
+package io.linkerd.mesh;
+
+/**
+ * Names are represented as hierarchical paths.
+ *
+ * A path element MAY contain binary-encoded segments.
+ */
+message Path {
+  repeated bytes elems = 1;
+}
+
+message PathNameTree {
+
+  /**
+   * A PathNameTree is a hierachy of nodes:
+   */
+  oneof node {
+    Nop nop = 1;
+    Alt alt = 2;
+    Union union = 3;
+    Leaf leaf = 4;
+  }
+
+  /**
+   * These states have no associated metadata and don't need the
+   * overhead of a message type.
+   */
+  enum Nop {
+    NEG = 0;
+    FAIL = 1;
+    EMPTY = 2;
+  }
+
+  /**
+   * Describes a sequence of alternate trees that may be used, from
+   * greatest preference to least.
+   */
+  message Alt { repeated PathNameTree trees = 1; }
+
+  /**
+   * Describes a set of trees to be used together.
+   */
+  message Union {
+    repeated Weighted trees = 1;
+    message Weighted {
+      double weight = 1;
+      PathNameTree tree = 2;
+    }
+  }
+
+  /**
+   * A terminal node with a value.
+   */
+  message Leaf { Path id = 1; }
+}
+
+/**
+ * A delegation table comprises a sequence of rewrite rules.  Each
+ * Dentry describes a prefix upon which paths are matched and a
+ * destination to use as the new prefix.
+ *
+ * These rules are applied until the tree cannot be refined any
+ * further.
+ */
+message Dtab {
+  repeated Dentry dentries = 1;
+
+  message Dentry {
+    Prefix prefix = 1;
+    PathNameTree dst = 2;
+
+    message Prefix {
+      repeated Elem elems = 1;
+
+      message Elem {
+        oneof value {
+          bytes label = 1;
+          Wildcard wildcard = 2;
+        }
+        message Wildcard {}
```

Comment:
i switched to message types everywhere.  enums were a premature optimization. i think it's better to use messages, so they may be extended in the future to support additional fields.